src/main/java/net/datafaker/internal/helper/COWMap.java

```diff
@@ -0,0 +1,93 @@
+package net.datafaker.internal.helper;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+public class COWMap<K, V> implements Map<K, V> {
+    private volatile Map<K, V> map;
+    private final Supplier<Map<K, V>> mapSupplier;
+
+    public COWMap(Supplier<Map<K, V>> mapSupplier) {
+        this.mapSupplier = mapSupplier;
+        this.map = mapSupplier.get();
+    }
+
+    @Override
+    public int size() {
+        return map.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return map.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        return map.containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        return map.containsValue(value);
+    }
+
+    @Override
+    public V get(Object key) {
+        return map.get(key);
+    }
+
+    @Override
+    public V put(K key, V value) {
+        Map<K, V> newMap = mapSupplier.get();
+        newMap.putAll(map);
+        final V result = newMap.put(key, value);
+        map = newMap;
+        return result;
+    }
+
+    @Override
+    public V remove(Object key) {
+        Map<K, V> newMap = mapSupplier.get();
+        newMap.putAll(map);
+        final V result = newMap.remove(key);
+        map = newMap;
```

Comment:
Isn't it better to keep the internal map as unmodifiable? Just for safety.
Like:
```java
map = Collections.unmodifiableMap(newMap);
```